### PR TITLE
Auto-detect upstream DNS when possible

### LINKED ISSUE #586: Automatically switch to local DNS when Google DNS blocked
A lot of people in corporate networks meet the DNS issue and it deters many newcomers from using Docksal.
It would be beneficial it we could automate the Upstream DNS switch.

On macOS and Linux current DNS server can be found in `/etc/resolv.conf`
On Windows it looks like Primary WINS Server might be the server we are looking for:

```bash
ipconfig /all | grep "Primary WINS"
```

We might also switch it back to `8.8.8.8` in case we detect that DOCKSAL UPSTREAM DNS is overridden but not accessible while `8.8.8.8` is accessible (person brought laptop home).

### PR DESCRIPTION
**Pull request by achekulaev**
Closes #586

FYI @sean-e-dietrich

### CODE REVIEW

File: bin/fin [hunk 1]
```diff
@@ -3231,16 +3231,57 @@ install_proxy_service ()
...
+			done
+		else
+			# In case no name servers found, fall back to the default
+			DOCKSAL_DNS_UPSTREAM="$DOCKSAL_DEFAULT_DNS"
+		fi
+	elif is_windows; then
+		# Look for Primary WINS server, it represents DNS whe WINS services is in use
+		local primary_wins=$(ipconfig /all | grep "Primary WINS")
+		if [[ "$primary_wins" =~ ([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}) ]]; then
+			DOCKSAL_DNS_UPSTREAM=${BASH_REMATCH[0]}
```
Comment (lmakarov) [inline bin/fin]: OMG what does that even do?? :)

File: bin/fin [hunk 2]
```diff
@@ -3231,16 +3231,57 @@ install_proxy_service ()
...
+					DOCKSAL_DNS_UPSTREAM="$server"
+				fi
+			done
+		else
+			# In case no name servers found, fall back to the default
+			DOCKSAL_DNS_UPSTREAM="$DOCKSAL_DEFAULT_DNS"
+		fi
+	elif is_windows; then
+		# Look for Primary WINS server, it represents DNS whe WINS services is in use
+		local primary_wins=$(ipconfig /all | grep "Primary WINS")
```
Comment (lmakarov) [inline bin/fin]: Can we rely on DNS and WINS being always the same server/IP on every Windows network?

File: bin/fin (hunk 1 above)
Comment (achekulaev) [reply bin/fin]: BASH_REMATCH array contains matches from the last regexp matching (line above).

File: bin/fin (hunk 2 above)
Comment (achekulaev) [reply bin/fin]: I can't be 100% sure of course. But it so for my local Windows 10 and for a guy in a strict corporate network with DNS banned and certificates-gated internet access, whom I asked to execute it https://github.com/docksal/docksal/issues/625#issuecomment-401441119

I am also asking Sean right now and will post update here.

File: bin/fin [hunk 3]
```diff
@@ -3231,16 +3231,57 @@ install_proxy_service ()
...
+
+	if is_linux || is_mac; then
+		# Get from /etc/resolv.conf
+		nameservers=$(grep "nameserver" "/etc/resolv.conf" 2>/dev/null | sed "s/nameserver //")
+		# Find first suitable
+		if [[ "$nameservers" != "" ]]; then
+			for server in $nameservers; do
+				# Check that it is an IP address and exclude Docksal IP
+				if [[ "$server" =~ [0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3} ]] &&
+					[[ "$server" != "192.168.64.100" ]]; then
```
[the remark this answers is not in the document]
Comment (achekulaev) [reply bin/fin]: Thanks for reminding. I refactored all those network variables. That was long overdue.

File: bin/fin (hunk 2 above)
Comment (achekulaev) [reply bin/fin]: Sean does not have Primary WINS in his configuration. So far so good. When WINS is present it is the DNS, which makes sense.

**Pull request merged**